wgpu/_compute.py

```diff
@@ -0,0 +1,152 @@
+"""
+Simple high-level utilities for doing compute on the GPU.
+
+The purpose of wgpu-py is to provide a Pythonic wrapper around wgpu-native.
+In principal, a higher-level API is not within the scope of the project.
+However, by providing a few un-opinionated utility functions, other projects
+(like python-shader) can use wgpu (e.g. in their tests) without having
+to keep track of changes in wgpu itself.
+"""
+
+import sys
+
+import wgpu
+
+# Note: these funcs need numpy, but we don't import it because its not a strict dep of wgpu-py
+
+
+def compute_with_buffers(input_arrays, output_arrays, shader, n=None):
+    """ Apply the given compute shader to the given input_arrays and return
+    output arrays. Both input and output arrays are represented on the GPU
+    using storage buffer objects.
+
+    Params:
+        input_arrays (dict): A dict mapping an integer binding to a numpy array.
+        output_arrays (dict): A dict mapping an integer binding to an outputs spec.
+            The spec must be a tuple ``(*shape, dtype)``. A buffer is created
+            for each binding and its data is read and returned. The binding can be
+            one that is also used as an input array (in which case the GPU buffer
+            is the same).
+        shader (bytes, shader-object): Any compatible shader object representing
+            a SpirV compute shader.
+        n (int, tuple, optional): The number provided to dispatch. Can be an int
+            or a 3-tuple of ints to specify (x, y, z). If not given or None,
+            the first value of the first spec in output_arrays is used.
+    """
+
+    # How can we get numpy arrays if numpy is not imported?
+    if "numpy" not in sys.modules:
+        raise RuntimeError("need numpy arrays.")
+    np = sys.modules["numpy"]
```

Comment:
This ain't pretty. Maybe it's a better idea to put utils like this in a submodule/subpackage that's not imported by default. But that depends a bit on the earlier comment about our intentions with this higher level API.

Reply:
Well, I think there's two concerns:

(1) Dependencies: if we want to make wgpu-py as broadly usable as possible, I think we should aim for being dependency free across the entire library, to avoid "gotcha" moments or even giving off the impression that we are tied to numpy and python-shader after all. So for example, instead of numpy, people could use the stdlib module `struct`, and instead of `python-shader`, people could `glsl2spirv` from the vulkan SDK. That should be the case throughout the library.

(2) Import time; Python imports are slow, and so we should try to keep importing wgpu-py fast. Probably it would be best to keep wgpu-py focused on a small core that is imported in one go, and keep other optional utilities separate, or opt-in.

Reply:
Fortunately, `python-shader` is not a dependency, you can pass in raw SPirV bytes too :)

Yeah, I think it would be better to not import this by default. I'm not sure about numpy, using struct or ctypes arrays could maybe work, but numpy is so much easier to work with. What if we did:
```py
try:
    import numpy as np
except ImportError:
    raise ImportError("wgpu does not need numpy, but this utility submodule does.")
```

Reply:
Well yes, numpy is convenient, but it's also a big dependency in terms of size on disk. Some people prefer to stick to things like struct to be able to ship lightweight software.